_posts/2022/2022-03-14-where-first-vs-find-by-know-what-your-scopes-are-doing.md

```diff
@@ -0,0 +1,80 @@
+---
+title: "Know what your scopes are doing: where.first vs find_by"
+description: "Attack of the abstracted SQL"
+layout: article
+category: ruby
+image:
+  base: "2022/where-first-vs-find-by-know-what-your-scopes-are-doing"
+  alt: "Wooden file card drawers"
+  credit: "Jan Antonin Kolar"
+  source: "https://unsplash.com/photos/lRoX0shwjUQ"
+
+---
+
+There are numerous benefits to using an Object-Relational Mapper (ORM), such as Active Record, to generate your SQL. You gain from clear repeatable abstractions, saving time and gaining readability.
+
+Unfortunately the very abstractions that make an ORM useful can also have unintended consequences.
+
+It is very easy to introduce performance problems because the _actual_ SQL that runs on your database server is generated by the framework.
+
+We recently came across an issue at CoverageBook that highlighted this very point.
+
+
+## Instead of:
+
+...using a `where` condition then `first`.
+
+```ruby
+User.where(email: "andy@goodscary.com").first
+```
+
+
+## Use:
+
+...`find_by`
+
+```ruby
+User.find_by(email: "andy@goodscary.com")
+User.find_by_email("andy@goodscary.com")
+```
+
+
+## Why?
+
+This is one of those cases where the ORM (and the tooling around it) get in the way and introduce unforseen performance issues.
+
+The `.where` scope has an implicit `ORDER` scope on the primary key that isn't obvious at first glance.
+
+```ruby
+User.where(email: "andy@goodscary.com")
+# SELECT "users".*
+# FROM "users"
+# WHERE "users"."email" = "andy@goodscary.com"
+
+User.where(email: "andy@goodscary.com").first
+# SELECT "users".*
+# FROM "users"
+# WHERE "users"."email" = "andy@goodscary.com"
+# ORDER BY "users"."id" ASC
+# LIMIT 1
+
+User.find_by(email: "andy@goodscary.com")
+# SELECT "users".*
+# FROM "users"
+# WHERE "users"."email" = "andy@goodscary.com"
+# LIMIT 1
+```
+
+Straightforward indexes on our database didn't help us as in our—more complex—case. We were querying using an index, but because we were using `.where().first` we were inadvertantly doing a non-indexed scan to establish the order which caused enormous performance problems.
```

Comment:
I know the way you've written it is grammatically correct...but (and, shocking I know...me going against what's grammatically correct) but sometimes I often leave a space in some cases with em-dashes because parsing can be difficult at times. For example, it took me a second to realise you weren't saying "our-more" and "complex-case". But this is not a request to change it. Just explaining how I first read it and why I don't always follow the grammatical rule here.

Reply:
Comma before 'which'.

Reply:
```suggestion
Straightforward indexes on our database didn't help us as in our—more complex—case. We were querying using an index, but because we were using `.where().first` we were inadvertantly doing a non-indexed scan to establish the order, which caused enormous performance problems.
```